### DISCUSSION TITLE
Is it possible to add media library filters?

### DISCUSSION THREAD

**Discussion by arjolpanci-nm** [area17/twill#1262] [Q&A] [answered]
For a particular use case of mine, it is very useful to have a filter which picks up all untagged media instances (all images which do not have a tag). However, the default filters can only filter by the particular tags, so this is not possible by default as far as I can see.

Is there a way to add a custom filter for the media library?

Thank you in advance.

**Comment by arjolpanci-nm** [comment]
I have found a workaround for this by listening to Media model created event using an observer and attaching a tag on it. I can later filter the requested media by using this tag.

This opened up another problem however. It seems that on media upload, twill has its own listeners which overrides the actions on my custom handler. This means that when I upload an image through the media library, it overrides the changes made by my "created" listener and it empties the tags before saving the model.

This only happens on uploads through the media library. I tested a manual upload implementation that I have on my application, and through there the event was handled properly and the media had the tag attached.

Is there a way to override this twill behaviour? Why are the tags emptied before the model is saved when uploading through the media library?

**Comment by pboivin** [comment, marked as answer]
Hi @arjolpanci-nm,

I'm not 100% sure but I think this could be what is deleting your custom tag:
https://github.com/area17/twill/blob/2.x/src/Repositories/Behaviors/HandleTags.php#L14-L17

I believe `afterSaveHandleTags()` is executed after your custom `created` event handler.

If you want to experiment with overriding this behavior on `MediaRepository`, one solution would be rebinding through Laravel's container:

```php
// file: app/Providers/AppServiceProvider.php

use A17\Twill\Repositories\MediaRepository;
use App\Repositories\AppMediaRepository;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    public function register()
    {
        $this->app->bind(MediaRepository::class, AppMediaRepository::class);
    }
    
    // ...
}
```

```php
// file: app/Repositories/AppMediaRepository.php

use A17\Twill\Repositories\MediaRepository;

class AppMediaRepository extends MediaRepository
{
    public function afterSave($object, $fields)
    {
        parent::afterSave($object, $fields);

        // attach custom tags
    }
    
    // OR override the behavior's method:

    public function afterSaveHandleTags($object, $fields)
    {
        // custom processing
    }
}
```

Hope this can help!

**Reply by arjolpanci-nm** [reply]
Yes, this did the trick!
Thank you.
